core/src/main/java/com/linecorp/armeria/internal/metric/RequestMetricSupport.java

```diff
@@ -40,56 +40,64 @@
+            AttributeKey.valueOf(RequestMetricSupport.class, "ACTIVE_REQUEST_METRICS");
 
-    private static final RequestMetrics PLACEHOLDER = new RequestMetricsPlaceholder();
+    private static final ActiveRequestMetrics PLACEHOLDER = new ActiveRequestMetricsPlaceHolder();
 
     public static void setup(RequestContext ctx, MeterIdPrefixFunction meterIdPrefixFunction) {
-        if (ctx.hasAttr(ATTR_REQUEST_METRICS)) {
+        if (ctx.hasAttr(ATTR_ACTIVE_REQUEST_METRICS)) {
             return;
         }
 
         // Set the attribute to the placeholder so that calling this method again has no effect.
         // The attribute will be set to the real one later in onRequestStart()
-        ctx.attr(ATTR_REQUEST_METRICS).set(PLACEHOLDER);
+        ctx.attr(ATTR_ACTIVE_REQUEST_METRICS).set(PLACEHOLDER);
 
         ctx.log().addListener(log -> onRequestStart(log, meterIdPrefixFunction),
                               RequestLogAvailability.REQUEST_HEADERS,
                               RequestLogAvailability.REQUEST_CONTENT);
         ctx.log().addListener(RequestMetricSupport::onRequestEnd,
                               RequestLogAvailability.REQUEST_END);
-        ctx.log().addListener(RequestMetricSupport::onResponse,
+        ctx.log().addListener(log -> onResponse(log, meterIdPrefixFunction),
                               RequestLogAvailability.COMPLETE);
     }
 
     private static void onRequestStart(RequestLog log, MeterIdPrefixFunction meterIdPrefixFunction) {
         final RequestContext ctx = log.context();
         final MeterRegistry registry = ctx.meterRegistry();
         final MeterIdPrefix idPrefix = meterIdPrefixFunction.apply(registry, log);
-        final RequestMetrics requestMetrics =  MicrometerUtil.register(
-                registry, idPrefix, RequestMetrics.class, DefaultRequestMetrics::new);
+        final MeterIdPrefix idPrefixActive = new MeterIdPrefix(idPrefix.name("activeRequests"),
+                                                               idPrefix.tags());
+
+        final ActiveRequestMetrics activeRequestMetrics =  MicrometerUtil.register(
+                registry, idPrefixActive, ActiveRequestMetrics.class, DefaultActiveRequestMetrics::new);
 
-        ctx.attr(ATTR_REQUEST_METRICS).set(requestMetrics);
-        requestMetrics.active().incrementAndGet();
+        ctx.attr(ATTR_ACTIVE_REQUEST_METRICS).set(activeRequestMetrics);
+        activeRequestMetrics.active().incrementAndGet();
     }
 
     private static void onRequestEnd(RequestLog log) {
-        final RequestMetrics metrics = requestMetrics(log);
-        metrics.requestDuration().record(log.requestDurationNanos(), TimeUnit.NANOSECONDS);
-        metrics.requestLength().record(log.requestLength());
         if (log.requestCause() != null) {
-            metrics.failure().increment();
-            metrics.active().decrementAndGet();
+            activeRequestMetrics(log).active().decrementAndGet();
```

Comment:
I'm inconfident if we really should decrement here given that `onResponse()` is expected to be always invoked.

A quick search shows that `endRequest(Throwable)` is always called with `endResponse(Throwable)` on the server side, which means we don't need to decrement here.

However, on the client side, `HttpRequestSubscriber.fail(Throwable)` only calls `endRequest(Throwable)`  without `endResponse(Throwable)`.

So, how about:

- Removing the decrement here and
- Adding a call to `endResponse(Throwable)` at `HttpRequestSubscriber.fail(Throwable)`

..?